enhancements/proxy-cache.md

```diff
@@ -0,0 +1,124 @@
+This proposal is to enable Quay to act as a pull through cache where, once pulled images are only pulled again when upstream images have been updated.   
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+
+## Open Questions
+
+ > 1. Will the check against the upstream registry count against an upstream rate limit?
+
+## Summary
+
+Dependencies on container images has increased tremendously with the adoption of container driven development. With the introduction of rate limits
+on popular container registries, Quay will act as a proxy cache to circumvent pull rate limitation from upstream registries. Adding a cache will also
+accelerate pull performance as images are pulled from cache rather than upstream dependencies. Cached images are updated only when the upstream
+image digest differs from cached image. 
+
+### Goals
+
+* A Quay user can define and configure(credentials, staleness period) via config file/app, a repository in Quay that acts a cache for a specific upstream registry.
+* A Quay super can user leverage storage quota of an organization to limit cache size. This means when cache size reaches its quota limit,
+  images from cache are evicted based on LRU.
+* A proxy cache organization will transparently cache and stream images to client. The images in the proxy cache organization should
+  support the default expected behaviour (like security scanner, time machine, etc) as other images on Quay.
+* Given the sensitive nature of accessing potentially untrusted upstream registry all cache pulls needs to be logged (audit log).
+* A Quay user can flush the cache to eliminate excess storage consumption.
+* Robot accounts can be created in the cache organization as usual, their RBAC can be managed for all existing cached repositories at a given point in time.
+* Provide metrics to track cache activity and efficiency (hit rate, size, evictions).
+
+### Non-Goals
+
+* In the first phase, configuring a cache proxy organization, caching upstreaming images, and quota management on cached repositories is the target.
+  Other goals will be implemented subsequently based on the work of this proposal.
+* Cached images are read-only.
+
+## Design Details
+
+The expected pull overview is depicted as below:
+![](https://user-images.githubusercontent.com/11522230/145866763-58f44c94-839b-4edb-a95b-b9c3648cf187.png)
+Design credits: @fmissi
+
+A user initiates a pulls of an image(say a `postgres:14` image) from an upstream repository on Quay. The repository is checked to see if the image is present.
+1. If the image does not exist, a fresh pull is initiated.
+  * The user is authenticated into the upstream registry and all the layers of `postgres:14` are pulled.
+  * The pulled layers are saved to cache and served to the user in parallel.
```

Comment:
Will serving and caching reuse the same connection to the upstream registry or will they open two different connections?

Reply:
I have not looked in depth at the technical feasibility of this, but my idea is to pull a layer from the upstream registry and then cache and stream it. So there should only be one pull against the upstream registry and we'd use the result of that to both cache and stream the layer to the user.